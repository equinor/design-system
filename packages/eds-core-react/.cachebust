# cache bust 1761833449
