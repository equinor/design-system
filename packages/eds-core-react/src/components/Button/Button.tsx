import { forwardRef, ButtonHTMLAttributes } from 'react'
import styled, { css, ThemeProvider } from 'styled-components'
import { token as buttonToken } from './tokens'
import { ButtonTokenSet, ButtonToken } from './Button.types'
import {
  typographyTemplate,
  bordersTemplate,
  outlineTemplate,
  spacingsTemplate,
  useToken,
  OverridableComponent,
} from '@equinor/eds-utils'
import { InnerFullWidth } from './InnerFullWidth'
import { useEds } from '../EdsProvider'

type Colors = 'primary' | 'secondary' | 'danger'
type Variants =
  | 'contained'
  | 'contained_icon'
  | 'outlined'
  | 'ghost'
  | 'ghost_icon'

const getVariant = (
  tokenSet: ButtonTokenSet,
  variant: Variants,
): ButtonToken => {
  switch (variant) {
    case 'ghost':
      return tokenSet.ghost
    case 'ghost_icon':
      return tokenSet.ghost_icon
    case 'outlined':
      return tokenSet.outlined
    case 'contained_icon':
      return tokenSet.contained_icon
    case 'contained':
    default:
      return tokenSet.contained
  }
}

const getToken = (variant: Variants, color: Colors): ButtonToken => {
  switch (color) {
    case 'danger':
      return getVariant(buttonToken.danger, variant)
    case 'secondary':
      return getVariant(buttonToken.secondary, variant)
    case 'primary':
    default:
      return getVariant(buttonToken.primary, variant)
  }
}

const Inner = styled.span`
  display: grid;
  grid-gap: var(--eds_button__gap, 8px);
  grid-auto-flow: column;
  align-items: center;
  height: 100%;
  justify-content: center;

  & > :is(svg, img) {
    margin-top: var(--eds_button__icon__margin_y, 0);
    margin-bottom: var(--eds_button__icon__margin_y, 0);
  }
`

const ButtonBase = styled.button(({ theme }: { theme: ButtonToken }) => {
  const { states, clickbound } = theme
  const { focus, hover, disabled } = states

  return css`
    box-sizing: border-box;
    margin: 0;
    padding: 0;
    text-decoration: none;
    position: relative;
    cursor: pointer;
    display: inline-block;
    background: ${theme.background};
    height: ${theme.height};
    width: ${theme.width};

    svg {
      justify-self: center;
    }

    ${spacingsTemplate(theme.spacings)}
    ${bordersTemplate(theme.border)}
    ${typographyTemplate(theme.typography)}

    &::before {
      position: absolute;
      top: 0;
      left: 0;
      width: auto;
      min-height: auto;
      content: '';
    }

    &::after {
      position: absolute;
      top: -${clickbound?.offset?.top};
      left: -${clickbound?.offset?.left};
      width: ${clickbound?.width};
      height: ${clickbound?.height};
      content: '';
    }

    @media (hover: hover) and (pointer: fine) {
      &:hover {
        background: ${hover.background};
        color: ${hover.typography?.color};
        ${bordersTemplate(hover?.border)};
      }
    }

    &:focus {
      outline: none;
    }

    &[data-focus-visible-added]:focus {
      ${outlineTemplate(focus.outline)}
    }
    &:focus-visible {
      ${outlineTemplate(focus.outline)}
    }
    /* Get rid of ff focus border for buttons */
    &::-moz-focus-inner {
      border: 0;
    }

    &:disabled,
    &[aria-disabled='true'] {
      cursor: not-allowed;
      background: ${disabled.background};
      ${bordersTemplate(disabled.border)};
      ${typographyTemplate(disabled.typography)};

      @media (hover: hover) and (pointer: fine) {
        &:hover {
          background: ${disabled.background};
        }
      }
    }
  `
})

export type ButtonProps = {
  /**  Specifies color */
  color?: Colors
  /** Specifies which variant to use */
  variant?: Variants
  /**
   * URL link destination
   * If defined, an 'a' element is used as root instead of 'button'
   */
  href?: string
  /** Is the button disabled */
  disabled?: boolean
  /** Type of button
   * @default 'button'
   */
  type?: string
  /** FullWidth (stretched) button  */
  fullWidth?: boolean
} & ButtonHTMLAttributes<HTMLButtonElement>

export const Button: OverridableComponent<ButtonProps, HTMLButtonElement> =
  forwardRef(function Button(
    {
      color = 'primary',
      variant = 'contained',
      children,
      disabled = false,
      href,
      tabIndex = 0,
      fullWidth = false,
      ...other
    },
    ref,
  ) {
    const { density } = useEds()
    const token = useToken({ density }, getToken(variant, color))

    const as = href && !disabled ? 'a' : other.as ? other.as : 'button'

    const type = href || other.as ? undefined : 'button'

    tabIndex = disabled ? -1 : tabIndex

    const buttonProps = {
      ref,
      as,
      href,
      type,
      disabled,
      tabIndex,
      ...other,
    }

    return (
      <ThemeProvider theme={token}>
        <ButtonBase {...buttonProps}>
          {fullWidth ? (
            <InnerFullWidth>{children}</InnerFullWidth>
          ) : (
            <Inner>{children}</Inner>
          )}
        </ButtonBase>
      </ThemeProvider>
    )
  })
