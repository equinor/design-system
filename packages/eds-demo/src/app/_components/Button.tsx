import React from "react";

type ButtonProps = {
  className?: string;
  icon?: React.ReactNode;
  onClick?: () => void;
  children?: React.ReactNode;
} & React.ButtonHTMLAttributes<HTMLButtonElement>;

export const PrimaryButton = ({
  className,
  icon,
  onClick,
  children = "Save",
  ...rest
}: ButtonProps) => {
  return (
    <button
      data-debug
      type="button"
      className={`select-none h-[36px] px-4 py-2.5 rounded border-opacity-0 justify-start gap-2 flex items-center button button--primary ${
        className || ""
      }`}
      onClick={onClick}
      {...rest}
    >
      {icon}
      <div className="text-sm leading-none">{children}</div>
    </button>
  );
};

export const SecondaryButton = ({
  className,
  icon,
  onClick,
  children = "Cancel",
  ...rest
}: ButtonProps) => {
  return (
    <button
      data-debug
      type="button"
      className={`select-none h-[36px] px-4 py-2.5 bg-opacity-0 rounded border justify-start gap-2 flex items-center ${
        className || ""
      }`}
      onClick={onClick}
      {...rest}
    >
      {icon}
      <div className="text-sm leading-none">{children}</div>
    </button>
  );
};
