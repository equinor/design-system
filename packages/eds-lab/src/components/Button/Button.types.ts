import type { ComponentToken } from '@equinor/eds-tokens'

export type ButtonToken = ComponentToken & {
  entities?: {
    icon: ComponentToken
  }
}

export type ButtonTokenSet = {
  contained: ButtonToken
  outlined: ButtonToken
  ghost: ButtonToken
  ghost_icon: ButtonToken
}
