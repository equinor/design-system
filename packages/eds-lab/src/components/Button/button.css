/* TODO update Figma-broker css generation with utility classes */

.eds-typography-navigation__btn {
  font-family: Equinor;
  font-size: 0.875rem;
  font-weight: 500;
  line-height: 1.143em;
}
/* TODO  */

/* Default is primary & contained styling */
.eds-btn {
  --eds-btn-color: var(--eds_text_static_icons__primary_white);
  --eds-btn-color--hover: var(--eds_text_static_icons__primary_white);
  --eds-btn-color--disabled: var(--eds_interactive_disabled__text);

  --eds-btn-background: var(--eds_interactive_primary__resting);
  --eds-btn-background--hover: var(--eds_interactive_primary__hover);
  --eds-btn-background--disabled: var(--eds_interactive_disabled__fill);

  --eds-btn-height: var(--eds_shape_button_min_height);
  --eds-btn-width: var(--eds_shape_button_min_width), inherit;

  --eds-btn-border-width: 1px;
  --eds-btn-border-color: var(--eds_interactive_primary__resting);
  --eds-btn-border-color--hover: var(--eds_interactive_primary__hover);
  --eds-btn-border-color--disabled: var(--eds_interactive_disabled__border);
  --eds-btn-border-radius: var(--eds_shape_button_border_radius);

  --eds-btn-clickbound-height: var(--eds_clickbound_default__base);
  --eds-btn-clickbound-top: -7px; /** calc didnt work :( */
  --eds-btn-clicbound-left: -0px;

  /** TODO make css token */
  --eds-btn-outline-color: rgba(0, 112, 121, 1);

  margin: 0;
  padding: 0;
  text-decoration: none;
  position: relative;
  cursor: pointer;
  display: inline-block;
  background: var(--eds-btn-background);
  height: var(--eds-btn-height);
  width: var(--eds-btn-width);
  padding-left: var(--eds_spacing_medium);
  padding-right: var(--eds_spacing_medium);

  border-width: var(--eds-btn-border-width);
  border-style: solid;
  border-color: var(--eds-btn-border-color);
  border-radius: var(--eds-btn-border-radius);

  color: var(--eds-btn-color);
  text-align: center;
}

.eds-btn::-moz-focus-inner {
  border: 0;
}

.eds-btn::before {
  position: absolute;
  top: 0;
  left: 0;
  width: auto;
  min-height: auto;
  content: '';
}

.eds-btn::after {
  position: absolute;
  top: var(--eds-btn-clickbound-top);
  left: var(--eds-btn-clicbound-left);
  width: 100%;
  height: var(--eds-btn-clickbound-height);
  content: '';
}

.eds-btn:hover {
  background: var(--eds-btn-background--hover);
  color: var(--eds-btn-color--hover);
  border-color: var(--eds-btn-border-color);
}

.eds-btn:focus {
  outline: none;
}

.eds-btn[data-focus-visible-added]:focus {
  outline-style: dashed;
  outline-offset: 2px;
  outline-width: 1px;
  outline-color: var(--eds-btn-outline-color);
}

.eds-btn:disabled {
  cursor: not-allowed;
  background: var(--eds-btn-background--disabled);
  border-color: var(--eds-btn-border-color--disabled);
  color: var(--eds-btn-color--disabled);
}

/** Colors */

.eds-btn.primary {
  --eds-btn-background: var(--eds_interactive_primary__resting);

  --eds-btn-color: var(--eds_interactive_primary__resting);
  --eds-btn-color--hover: var(--eds_interactive_primary__resting);
  --eds-btn-background--hover: var(--eds_interactive_primary__hover_alt);
  --eds-btn-border-color: var(--eds_interactive_primary__resting);
  --eds-btn-border-color--hover: var(--eds_interactive_primary__hover);
}
.eds-btn.secondary {
  --eds-btn-background: var(--eds_interactive_secondary__resting);

  --eds-btn-color: var(--eds_interactive_secondary__resting);
  --eds-btn-color--hover: var(--eds_interactive_secondary__resting);
  --eds-btn-background--hover: var(--eds_interactive_secondary__highlight);
  --eds-btn-border-color: var(--eds_interactive_secondary__resting);
  --eds-btn-border-color--hover: var(--eds_interactive_secondary__link_hover);
}
.eds-btn.danger {
  --eds-btn-background: var(--eds_interactive_danger__resting);

  --eds-btn-color: var(--eds_interactive_danger__resting);
  --eds-btn-color--hover: var(--eds_interactive_danger__resting);
  --eds-btn-background--hover: var(--eds_interactive_danger__highlight);
  --eds-btn-border-color: var(--eds_interactive_danger__resting);
  --eds-btn-border-color--hover: var(--eds_interactive_danger__hover);
}

/** Variants */
.eds-btn.contained {
  --eds-btn-color: var(--eds_text_static_icons__primary_white);
  --eds-btn-color--hover: var(--eds_text_static_icons__primary_white);
}
.eds-btn.contained.primary {
  --eds-btn-background--hover: var(--eds_interactive_primary__hover);
  --eds-btn-border-color--hover: var(--eds_interactive_primary__hover);
}
.eds-btn.contained.secondary {
  --eds-btn-background--hover: var(--eds_interactive_secondary__link_hover);
  --eds-btn-border-color--hover: var(--eds_interactive_secondary__link_hover);
}
.eds-btn.contained.danger {
  --eds-btn-background--hover: var(--eds_interactive_danger__hover);
  --eds-btn-border-color--hover: var(--eds_interactive_danger__hover);
}

.eds-btn.outlined {
  --eds-btn-background: transparent;
  --eds-btn-background--disabled: transparent;
}

.eds-btn.ghost {
  --eds-btn-background: transparent;
  --eds-btn-background--disabled: transparent;
  --eds-btn-border-width: 0;
}

.eds-btn.ghost_icon {
  --eds-btn-background: transparent;
  --eds-btn-background--disabled: transparent;

  --eds-btn-height: var(--eds_shape_icon_button_min_height);
  --eds-btn-width: var(--eds_shape_icon_button_min_width);

  --eds-btn-border-width: 0;
  --eds-btn-border-radius: 50%;
}

.eds-btn.compact {
  /** TODO add compact shape css tokens */
  --eds-btn-height: 24px;
  --eds-btn-clickbound-height: var(--eds_clickbound_compact__standard);
  --eds-btn-clickbound-top: -5px;
}

.eds-btn.ghost_icon.compact {
  /** TODO add compact shape css tokens */
  --eds-btn-height: 32px;
  --eds-btn-width: 32px;
  --eds-btn-clickbound-top: 0px;
}

.eds-btn-inner {
  display: grid;
  grid-gap: 8px;
  grid-auto-flow: column;
  align-items: center;
  height: 100%;
  justify-content: center;
}

.eds-btn-inner.full-width {
  height: 100%;
  display: flex;
  align-items: center;
}

.eds-btn-inner.full-width > img:first-child,
.eds-btn-inner.full-width > svg:first-child {
  margin-right: 8px;
}

.eds-btn-inner.full-width > img:last-child,
.eds-btn-inner.full-width > svg:last-child {
  margin-left: 8px;
}

.eds-btn-inner.full-width > img:only-child,
.eds-btn-inner.full-width > svg:only-child {
  margin-left: auto;
  margin-right: auto;
}

.eds-btn-inner.full-width > span:first-child {
  margin-left: 32px;
}

.eds-btn-inner.full-width > span:last-child {
  margin-right: 32px;
}

.eds-btn-inner.full-width > span:only-child {
  margin-right: 0;
  margin-left: 0;
}
