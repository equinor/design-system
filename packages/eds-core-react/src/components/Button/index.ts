import { Button as ButtonWrapper, ButtonProps } from './Button'
import { ButtonGroup, ButtonGroupProps } from './ButtonGroup/ButtonGroup'
import { ToggleButton, ToggleButtonProps } from './ToggleButton/ToggleButton'

type ButtonCompoundProps = typeof ButtonWrapper & {
  Group: typeof ButtonGroup
  Toggle: typeof ToggleButton
}

const Button = ButtonWrapper as ButtonCompoundProps
Button.Group = ButtonGroup
Button.Toggle = ToggleButton

Button.Group.displayName = 'Button.Group'
Button.Toggle.displayName = 'Button.Toggle'

export { Button }
export type { ButtonProps, ButtonGroupProps, ToggleButtonProps }
