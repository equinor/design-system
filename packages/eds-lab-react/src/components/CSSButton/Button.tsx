import {
  forwardRef,
  ElementType,
  ButtonHTMLAttributes,
  createElement,
} from 'react'
import { clsx } from 'clsx'
import { InnerFullWidth } from './InnerFullWidth'
import { useEds } from '@equinor/eds-core-react'
import './button.css'
//** Injecting root css properties needs to be solved */
import '@equinor/eds-tokens/tokens.css'

export type ButtonProps = {
  /**  Specifies color */
  color?: 'primary' | 'secondary' | 'danger'
  /** Specifies which variant to use */
  variant?: 'contained' | 'outlined' | 'ghost' | 'ghost_icon'
  /**
   * URL link destination
   * If defined, an 'a' element is used as root instead of 'button'
   */
  href?: string
  /** Is the button disabled */
  disabled?: boolean
  /** Change html element. */
  as?: ElementType
  /** FullWidth (stretched) button  */
  fullWidth?: boolean
} & ButtonHTMLAttributes<HTMLButtonElement>

export const CSSButton = forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    {
      color = 'primary',
      variant = 'contained',
      children,
      disabled = false,
      href,
      tabIndex = 0,
      fullWidth = false,
      className,
      ...other
    },
    ref,
  ) {
    const { density } = useEds()

    const as: ElementType =
      href && !disabled ? 'a' : other.as ? other.as : 'button'
    // const type = href || other.as ? undefined : 'button'

    const classNames: string = clsx([
      'eds-typography-navigation__btn',
      'eds-btn',
      color,
      variant,
      density,
      className,
    ])

    const buttonProps = {
      ref,
      href,
      disabled,
      tabIndex,
      className: classNames,
      ...other,
    }

    return createElement(
      as,
      buttonProps,
      fullWidth ? (
        <InnerFullWidth>{children}</InnerFullWidth>
      ) : (
        <span className="eds-btn-inner">{children}</span>
      ),
    )
  },
)
