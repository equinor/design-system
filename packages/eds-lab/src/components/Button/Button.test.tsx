/* eslint-disable no-undef */
import { render, cleanup, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'jest-styled-components'
import styled from 'styled-components'
import { Icon } from '@equinor/eds-core-react'
import { save } from '@equinor/eds-icons'
import { Button } from './Button'
import React from 'react'

Icon.add({ save })

const StyledButton = styled(Button)`
  position: relative;
  height: 100px;
  width: 100px;
`

const MarginButton = styled(Button)`
  margin: 12px;
`

afterEach(cleanup)

describe('Button', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(<Button>Button</Button>)
    expect(asFragment()).toMatchSnapshot()
  })
  it('Has provided icon when variant is icon', () => {
    render(
      <Button variant="ghost_icon">
        <Icon name="save" title="save me test" />
      </Button>,
    )
    expect(screen.getByLabelText('save me test')).toBeInTheDocument()
    expect(screen.queryByTestId('eds-icon-path')).toHaveAttribute(
      'd',
      save.svgPathData,
    )
  })
  it('Can extend the css for the component', () => {
    render(<StyledButton>Test me!</StyledButton>)
    const button = screen.getByRole('button')

    expect(button).toHaveStyleRule('position', 'relative')
    expect(button).toHaveStyleRule('height', '100px')
    expect(button).toHaveStyleRule('width', '100px')
  })
  it('onSubmit is called one time when used in form', () => {
    const onSubmit = jest.fn((e: React.FormEvent<HTMLFormElement>) =>
      e.preventDefault(),
    )

    render(
      <form onSubmit={onSubmit}>
        <Button type="submit">Submit button</Button>
      </form>,
    )

    const submitButton = screen.queryByText('Submit button')

    fireEvent.submit(submitButton)

    expect(onSubmit).toHaveBeenCalledTimes(1)
  })

  // TODO Why do we want this behaviour?
  // it('onSubmit is ignored in form by default', () => {
  //   const onSubmit = jest.fn((e: React.FormEvent<HTMLFormElement>) =>
  //     e.preventDefault(),
  //   )

  //   render(
  //     <form onSubmit={onSubmit}>
  //       <Button>Submit button</Button>
  //     </form>,
  //   )

  //   const submitButton = screen.queryByText('Submit button')

  //   fireEvent.click(submitButton)

  //   expect(onSubmit).toHaveBeenCalledTimes(0)
  // })
  it('Has provided icon when icon and text is defined', () => {
    render(
      <Button>
        <Icon name="save" title="save"></Icon>Button
      </Button>,
    )
    expect(screen.getByLabelText('save')).toBeInTheDocument()
    expect(screen.queryByTestId('eds-icon-path')).toHaveAttribute(
      'd',
      save.svgPathData,
    )
  })
  it('Renders as a link when href prop is defined', () => {
    render(<Button href="/" />)
    const button = screen.getByRole('link')

    expect(button).toBeInTheDocument()
  })
  it('Can change margins', () => {
    render(<MarginButton>Test me!</MarginButton>)
    const button = screen.getByRole('button')
    expect(button).toHaveStyleRule('margin', '12px')
  })
})
