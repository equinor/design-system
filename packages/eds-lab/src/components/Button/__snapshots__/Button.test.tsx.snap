// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="eds-typography-navigation__btn eds-btn primary contained comfortable"
    tabindex="0"
  >
    <span
      class="eds-btn-inner"
    >
      Button
    </span>
  </button>
</DocumentFragment>
`;
